sim/adversary/spam.go

```diff
@@ -0,0 +1,87 @@
+package adversary
+
+import (
+	"github.com/filecoin-project/go-f3/gpbft"
+)
+
+var _ Receiver = (*Spam)(nil)
+
+// Spam is an adversary that propagates COMMIT messages for bottom for a
+// configured number of future rounds.
+type Spam struct {
+	id          gpbft.ActorID
+	host        Host
+	roundsAhead uint64
+}
+
+// NewSpam instantiates a new Spam adversary that spams the network with
+// spammable messages (i.e. COMMIT for bottom) for the configured number of
+// roundsAhead via either synchronous or regular broadcast. This adversary
+// resigns the spammable messages as its own to mimic messages with valid
+// signature but for future rounds.
+func NewSpam(id gpbft.ActorID, host Host, roundsAhead uint64) *Spam {
+	return &Spam{
+		id:          id,
+		host:        host,
+		roundsAhead: roundsAhead,
+	}
+}
+
+func NewSpamGenerator(power *gpbft.StoragePower, roundsAhead uint64) Generator {
+	return func(id gpbft.ActorID, host Host) *Adversary {
+		return &Adversary{
+			Receiver: NewSpam(id, host, roundsAhead),
+			Power:    power,
+		}
+	}
+}
+
+func (r *Spam) ReceiveMessage(msg *gpbft.GMessage, _ bool) (bool, error) {
+	// Limit spammer to the spammable message, that is COMMIT message for bottom
+	// which has no justification. QUALITY and PREPARE message without justification
+	// are only valid for round zero and get filtered out by validation if spammed.
+	// So we are left with COMMITs with no justification as the spam attack vector.
+	if msg.Vote.Step == gpbft.COMMIT_PHASE && msg.Vote.Value.IsZero() {
+
+		// Spam the network with COMMIT messages by incrementing rounds up to
+		// roundsAhead.
+		for spamRound := uint64(0); spamRound < r.roundsAhead; spamRound++ {
+			spamVote := gpbft.Payload{
+				Instance: msg.Vote.Instance,
+				Round:    msg.Vote.Round + spamRound,
+				Step:     msg.Vote.Step,
+				Value:    msg.Vote.Value,
+			}
+			sigPayload := r.host.MarshalPayloadForSigning(&spamVote)
+			_, power, beacon := r.host.GetCanonicalChain()
+			_, pubkey := power.Get(r.id)
+
+			sig, err := r.host.Sign(pubkey, sigPayload)
+			if err != nil {
+				panic(err)
+			}
+
+			var ticket gpbft.Ticket
+			if len(msg.Ticket) != 0 {
+				var err error
+				ticket, err = gpbft.MakeTicket(beacon, spamVote.Instance, spamVote.Round, pubkey, r.host)
+				if err != nil {
+					panic(err)
+				}
+			}
+			r.host.Broadcast(&gpbft.GMessage{
+				Sender:    r.ID(),
+				Vote:      spamVote,
+				Signature: sig,
+				Ticket:    ticket,
+			})
+		}
+	}
+	return true, nil
+}
+
+func (r *Spam) ID() gpbft.ActorID                                              { return r.id }
+func (r *Spam) Start() error                                                   { return nil }
```

Comment:
I'm a bit confused why this adversary waits to receive a COMMIT bottom from someone else (which may never happen) instead of just immediately sending the spam when started. You just need to know the instance number then you can generate as much spam as wanted right away. There's no need to derive it from other messages. (Generating more later on isn't a bad idea, but you could just set an alarm for that, or look at the clock when receiving any message).